### PR TITLE
Add URI SANs

### PR DESCRIPTION
**Pull request by jefferai**
This is a re-push of #4675 so that it can actually be dev'd in this repo.

Obsoletes #4675

### CODE REVIEW

File: builtin/logical/pki/cert_util.go [hunk 1]
```diff
@@ -1537,24 +1607,34 @@ func handleOtherSANs(in *x509.Certificate, sans map[string][]string) error {
 	return nil
 }
 
+const (
+	nameTypeEmail = 1
+	nameTypeDNS   = 2
+	nameTypeURI   = 6
+	nameTypeIP    = 8
```
Comment (briankassouf) [inline builtin/logical/pki/cert_util.go]: Previously this was `7`, should it still be?

File: builtin/logical/pki/cert_util.go [hunk 2]
```diff
@@ -1537,24 +1607,34 @@ func handleOtherSANs(in *x509.Certificate, sans map[string][]string) error {
 	return nil
 }
 
+const (
+	nameTypeEmail = 1
```
Comment (briankassouf) [inline builtin/logical/pki/cert_util.go]: Could you add a comment explaining where these values are coming from, with a link?

Comment (jefferai) [reply builtin/logical/pki/cert_util.go]: Fixed.

File: builtin/logical/pki/cert_util.go (hunk 1 above)
Comment (jefferai) [reply builtin/logical/pki/cert_util.go]: Fixed.

File: builtin/logical/pki/cert_util.go [hunk 3]
```diff
@@ -861,22 +860,87 @@ func generateCreationBundle(b *backend, data *dataBundle) error {
...
+					if !valid {
+						return errutil.UserError{Err: fmt.Sprintf(
+							"URI Subject Alternative Names were provided via CSR which are not valid for this role"),
+						}
+					}
+
+					parsedURI, err := url.Parse(uri)
+					if parsedURI == nil || err != nil {
+						return errutil.UserError{Err: fmt.Sprintf(
+							"the provided URI Subject Alternative Name '%s' is not a valid URI", uri),
```
[the remark this answers is not in the document]
Comment (jefferai) [reply builtin/logical/pki/cert_util.go]: Convention is errors are returned lowercase (the ones above start with acronyms)

**Pull request merged**